src/main/java/org/apache/sling/feature/cpconverter/handlers/BundleEntryHandler.java

```diff
@@ -108,111 +144,302 @@ public void handle(@NotNull String path, @NotNull Archive archive, @NotNull Entr
+        
+        if (StringUtils.isNotBlank(typesHeader) ) {
+            for (ManifestHeader.Entry entry : ManifestHeader.parse(typesHeader).getEntries()) {
+                JarEntry jarEntry = jarFile.getJarEntry(entry.getValue());
+                if (jarEntry == null) {
+                    logger.warn("createNamespaceRegistry: Bundle {} has referenced a non existing node type definition: {}",
+                            manifest.getMainAttributes().getValue(Constants.BUNDLE_SYMBOLICNAME), entry.getValue());
+                } else {
+                    try (InputStream inputStream = jarFile.getInputStream(jarEntry);
+                         Reader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8)) {
+                        registry.registerCnd(reader, entry.getValue());
                     }
                 }
             }
         }
+        return registry;
     }
 
-    // method visibility set to 'protected' fot testing purposes
-    protected @NotNull Properties readGav(@NotNull String entryName, @NotNull JarInputStream jarInput) throws IOException {
-        Properties pomProperties = new Properties();
-        Properties pomXmlProperties = new Properties();
+    /**
+     * Lazily initializes the cache with the necessary VaultPackageAssemblers
+     * @param bundleArtifactId
+     * @param repositoryPath
+     * @param cache
+     * @param converter
+     * @return the VaultPackageAssembler from the cache to use for the given repository path
+     */
+    public VaultPackageAssembler initPackageAssemblerForPath(@NotNull ArtifactId bundleArtifactId, @NotNull String repositoryPath, @NotNull PathEntry pathEntry, @NotNull Map<PackageType, VaultPackageAssembler> cache, @NotNull ContentPackage2FeatureModelConverter converter) {
+        PackageType packageType = VaultPackageUtils.detectPackageType(repositoryPath);
+        VaultPackageAssembler assembler = cache.get(packageType);
+        if (assembler == null) {
+            final String packageNameSuffix;
+            switch (packageType) {
+                case APPLICATION:
+                    packageNameSuffix = "-apps";
+                    break;
+                case CONTENT:
+                    packageNameSuffix = "-content";
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected package type " + packageType + " detected for path " + repositoryPath);
+            }
+            final PackageId packageId = new PackageId(bundleArtifactId.getGroupId(), bundleArtifactId.getArtifactId()+packageNameSuffix, bundleArtifactId.getVersion());
+            assembler = VaultPackageAssembler.create(converter.getTempDirectory(), packageId, "Generated out of Sling Initial Content from bundle " + bundleArtifactId + " by cp2fm");
+            cache.put(packageType, assembler);
+        }
+        
+        DefaultWorkspaceFilter filter = assembler.getFilter();
```

Comment:
@stefanseifert I need some support here. Unsure which `ImportMode` matches the different flags of the JCR Contentloader best (compare also with [SLING-10318](https://issues.apache.org/jira/browse/SLING-10318))

Reply:
i'm not sure if anyone using sling-initial-content for deploying application content to /apps is using anything different then "overwrite" (we do not). the other variants like overwriteProperties, merge etc. sound more useful for things like importing content. probably we can restrict complexity here by only supporting a subset first (e.g. overwrite) and fail with an descriptive error message if anything different is used.

Reply:
This extraction creates both application and content packages. But even `overwrite` behaves differently for JSON and JCR XML IMHO. I would also appreciate a comment in https://issues.apache.org/jira/browse/SLING-10318